pkg/mover/intermediate.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/internal"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+const (
+	defaultPerm fs.FileMode = 0644
+)
+
+// saveIntermediateBundle will tar in this order:
+// - The original chart
+// - The hits file
+// - The container images detected as references in the chart
+//
+// The hints file goes first in the tar, followed by the chart files.
+// Finally, images are appended using the go-containerregistry tarball lib
+func saveIntermediateBundle(cd *ChartData, tarFile string, log Logger) error {
+	tmpTarball, err := os.CreateTemp("", "intermediate-bundle-tar-*")
+	if err != nil {
+		return err
+	}
+	tmpTarballFilename := tmpTarball.Name()
+	tfw := wrapAsTarFileWriter(tmpTarball)
+
+	// hints file goes first to be extracted quickly on demand
+	log.Printf("Writing %s...\n", IntermediateBundleHintsFilename)
+	if err := tfw.WriteMemoryFile(IntermediateBundleHintsFilename, cd.rawHints, defaultPerm); err != nil {
+		return fmt.Errorf("failed to write %s: %w", IntermediateBundleHintsFilename, err)
+	}
+
+	log.Printf("Writing Helm Chart files at %s/...\n", cd.chart.Metadata.Name)
+	if err := tarChart(tfw, cd.chart); err != nil {
+		return fmt.Errorf("failed archiving original-chart/: %w", err)
+	}
+
+	if err := packImages(tfw, cd.imageChanges, log); err != nil {
+		return fmt.Errorf("failed archiving images: %w", err)
+	}
+
+	if err := tfw.Close(); err != nil {
+		return fmt.Errorf("failed closing intermediate bundle temp file %s: %w",
+			tmpTarballFilename, err)
+	}
+
+	if err := copyAs(tmpTarballFilename, tarFile); err != nil {
+		return fmt.Errorf("failed copying  %s -> %s: %w", tmpTarballFilename, tarFile, err)
+	}
+	log.Printf("Intermediate bundle complete at %s\n", tarFile)
+	return nil
+}
+
+func copyAs(src, dst string) error {
+	tmpFilename := fmt.Sprintf("%s.tmp", dst)
+	dstFile, err := os.Create(tmpFilename)
+	if err != nil {
+		return fmt.Errorf("failed to create temporary destination file %s: %w", tmpFilename, err)
+	}
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open source %s: %w", src, err)
+	}
+	if _, err := io.Copy(dstFile, srcFile); err != nil {
+		return fmt.Errorf("failed copying %s -> %s: %w", src, dst, err)
+	}
+	if err := os.Rename(tmpFilename, dst); err != nil {
+		return fmt.Errorf("failed renaming %s -> %s: %w", tmpFilename, dst, err)
+	}
+	return nil
+}
+
+// tarChart tars all files from the original chart into `original-chart/`
+func tarChart(tfw *tarFileWriter, chart *chart.Chart) error {
+	for _, file := range chart.Raw {
+		if err := tfw.WriteMemoryFile(filepath.Join("original-chart", file.Name), file.Data, defaultPerm); err != nil {
+			return fmt.Errorf("failed to write chart's inner file %s: %v", file.Name, err)
+		}
+	}
+	return nil
+}
+
+func packImages(tfw *tarFileWriter, imageChanges []*internal.ImageChange, logger Logger) error {
+	root := "/"
+	fsys := os.DirFS(root)
```

Comment:
remove the filesystem

Reply:
now I save the stream with a name and size